File: 프로그래머스/lv2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    vector<vector<int>> prefix(discount.size() + 1);
    for (int i = 0; i < discount.size() + 1; i++)
        prefix[i].resize(number.size());
    for (int i = 0; i < number.size(); i++)
        prefix[0][i] = 0;
    for (int i = 0; i < discount.size(); i++) {
        for (int j = 0; j < number.size(); j++) {
            prefix[i+1][j] = prefix[i][j] + (want[j] == discount[i]);
        }
    }
    for (int i = 0; i < discount.size() - 9; i++) {
        bool flag = true;
        for (int j = 0; j < number.size(); j++) {
            if (prefix[i+10][j] - prefix[i][j] < number[j]) {
                flag = false;
                break;
            }
        }
        if (flag) answer++;
    }
    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < 10; i++) {
            if (map.containsKey(discount[i]))
                map.put(discount[i], map.get(discount[i]) + 1);
            else map.put(discount[i], 1);
        }
        int cnt = 1;
        for (int i = 0; i < want.length; i++) {
            if (!map.containsKey(want[i]) || number[i] != map.get(want[i])) {
                cnt--;
                break;
            }
        }
        for (int i = 10; i < discount.length; i++, cnt++) {
            map.put(discount[i-10], map.get(discount[i-10]) - 1);
            if (map.containsKey(discount[i]))
                map.put(discount[i], map.get(discount[i]) + 1);
            else map.put(discount[i], 1);
            for (int j = 0; j < want.length; j++) {
                if (!map.containsKey(want[j]) || number[j] != map.get(want[j])) {
                    cnt--;
                    break;
                }
            }
        }
        return cnt;
    }
}